test-cloud/test-cloud-performance/src/test/java/org/kie/cloud/integrationtests/s2i/KieServerS2iJbpmEJBTimerPerfIntegrationTest.java

```diff
@@ -0,0 +1,358 @@
+
+        return fireAtTime;
+    }
+    
+    protected void startAndWaitForStartingThreads(int numberOfThreads, Duration duration, Integer iterations, Runnable runnable) {
+       List<Thread> threads = new ArrayList<>();
+        for (int i = 0; i < numberOfThreads; i++) {
+            Thread t = new Thread(new RunnableWrapper(duration, iterations, runnable));
+            t.start();
+            threads.add(t);
+        }
+
+        threads.forEach(thread -> {
+            try {
+                thread.join();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+        
+    }
+    
+    protected Runnable getStartingRunnable(String containerId, String processId, Map<String, Object> parameters) {
+        return () -> {
+            try {
+                long pid = processServicesClient.startProcess(containerId, processId, parameters);
+                assertThat(pid).isNotNull();
+            } catch (KieServicesHttpException e) {
+                logger.error("There has been an error while starting processes", e);
+                throw e;
+            }
+        };
+    }
+    
+    protected void waitForAllProcessesToComplete(Duration waitForCompletionDuration) {
+        BooleanSupplier completionCondition = () -> queryServicesClient.findProcessInstancesByStatus(ACTIVE_STATUS, 0, 1).isEmpty();
+        TimeUtils.wait(waitForCompletionDuration, Duration.of(1, ChronoUnit.SECONDS), completionCondition);
+    }
+    
+    private void gatherAndAssertStatistics() {
+        int numberOfPages = 1 + (PROCESSES_COUNT / 5000);// including one additional page to check there are no more processes
+
+        List<ProcessInstance> completedProcesses = new ArrayList<>(PROCESSES_COUNT);
+
+        for (int i = 0; i < numberOfPages; i++) {
+            List<ProcessInstance> response = queryServicesClient.findProcessInstancesByStatus(Collections.singletonList(org.jbpm.process.instance.ProcessInstance.STATE_COMPLETED), i, 5000);
+            completedProcesses.addAll(response);
+        }
+
+        logger.info("Completed processes count: {}", completedProcesses.size());
+        
+        List<ProcessInstance> activeProcesses = queryServicesClient.findProcessInstancesByStatus(Collections.singletonList(org.jbpm.process.instance.ProcessInstance.STATE_ACTIVE), 0, 100);
+        logger.info("Active processes count: {}", activeProcesses.size());
+        
+        assertThat(activeProcesses).isEmpty();
+        
+        assertThat(completedProcesses).hasSize(PROCESSES_COUNT);
+       
+        Map<String, Integer> startedHostNameDistribution = new HashMap<>();
+        completedHostNameDistribution = new HashMap<>();
+       
+        for (String pod : pods) {
+            completedHostNameDistribution.put(pod, 0);
+            startedHostNameDistribution.put(pod, 0);
+        }
+        
+        for (int i = 0; i < numberOfPages-1; i++) {
```

Comment:
I assume this is because multiplications of 5000 will query here the non-existent page, which is added because we want to first check if there is really for example 10k processes and not more.

However, this -1 will not work for example for 14 000 processes. 14 000/5000 will return 2, and +1 is 3 on line 300. So here you will only go through pages 0 and 1. I suggest that for first check if there are as many processes completed as we want, we should use for example:
```
int numberOfPages = ((Double)Math.ceil(PROCESSES_COUNT / 5000)).intValue(); // number of pages for normal query
```
And then in the first check we just add +1 to query one more page.

Or we can leave it as it is now and don't decrease numberOfPages by 1 on line 326, because it should have no effect if we want page which does not exist - it should give us 0 results, in fact, we do it like that in the first check (line 304-3007) and it worked as far as I know.

Reply:
Removed the -1 on line 326 as you suggests as second option